### PR TITLE
[android] Refactor mapview + associated classes to high-level JNI bindings

### LINKED ISSUE #6533: JNI cleanup/refactor
We need to find a way to make our current `jni.cpp` implementation a bit more manageable. 
Using [jni.hpp](https://github.com/mapbox/jni.hpp) has given us a modern type safe wrapper c++14 approach to the jni but it doesn't help us manage the amount of configuration we currently need to do. We should look into structuring the content differently or approach the idea in a different way.

cc @mapbox/android

### PR DESCRIPTION
**Pull request by ivovandongen**
Fixes #6533

Main classes:
- NativeMapView
- FileSource
- Offline

### CODE REVIEW

**Comment by tobrun** [comment]
Looking great so far!

**Comment by tobrun** [comment]
I'm able to build `arm-v7` but hitting build issues for `arm-v8`

**Comment by ivovandongen** [comment]
@tobrun Could you clean your build first and try again? Just did a fresh build without problems:

```
17:24 $ make android-lib-arm-v8
platform/android/scripts/ndk.sh arm-v8 arm64-v8a 21 > build/android-arm-v8-21/Debug/env.sh.tmp && mv build/android-arm-v8-21/Debug/env.sh.tmp build/android-arm-v8-21/Debug/env.sh
Using system-provided Android NDK at /Users/ivo/Library/Android/sdk/ndk-bundle
# Invoke CMake twice to fix issues from double inclusion of toolchain.cmake on the first run.
. build/android-arm-v8-21/Debug/env.sh && ([ -f build/android-arm-v8-21/Debug/build.ninja ] || ${CMAKE} -H. -B"build/android-arm-v8-21/Debug" -G"${CMAKE_GENERATOR}" ${CMAKE_ARGS} -DCMAKE_BUILD_TYPE=Debug -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DMBGL_PLATFORM=android -DMASON_PLATFORM=android -DMASON_PLATFORM_VERSION=arm-v8-21) && ${CMAKE} -H. -B"build/android-arm-v8-21/Debug"
-- Check for working CXX compiler: /Users/ivo/Library/Android/sdk/ndk-bundle/toolchains/llvm/prebuilt/darwin-x86_64/bin/clang++
-- Check for working CXX compiler: /Users/ivo/Library/Android/sdk/ndk-bundle/toolchains/llvm/prebuilt/darwin-x86_64/bin/clang++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Check for working C compiler: /Users/ivo/Library/Android/sdk/ndk-bundle/toolchains/llvm/prebuilt/darwin-x86_64/bin/clang
-- Check for working C compiler: /Users/ivo/Library/Android/sdk/ndk-bundle/toolchains/llvm/prebuilt/darwin-x86_64/bin/clang -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Updating submodules...
[Mason] Downloading package https://mason-binaries.s3.amazonaws.com/android-arm-v8-21/gtest/1.8.0.tar.gz...
[Mason] Unpacking package to mason_packages/android-arm-v8-21/gtest/1.8.0...
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/ivo/git/mapbox-gl-native/build/android-arm-v8-21/Debug
-- Updating submodules...
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/ivo/git/mapbox-gl-native/build/android-arm-v8-21/Debug
. build/android-arm-v8-21/Debug/env.sh && ${CMAKE} --build build/android-arm-v8-21/Debug --  -j8 mapbox-gl example-custom-layer
[230/230] Linking CXX shared library libexample-custom-layer.so
```

File: platform/android/src/geometry/feature.cpp [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+namespace android {
+
+jni::Object<Feature> Feature::fromGeometry(jni::JNIEnv& env, jni::Object<Geometry> geometry, jni::Object<JsonObject> properties, jni::String id) {
+    static auto method = Feature::javaClass.GetStaticMethod<jni::Object<Feature> (jni::Object<Geometry>, jni::Object<JsonObject>, jni::String)>(env, "fromGeometry");
+    return Feature::javaClass.Call(env, method, geometry, properties, id);
+}
+
+void Feature::registerNative(jni::JNIEnv& env) {
+    // Lookup the class
+    Feature::javaClass = *jni::Class<Feature>::Find(env).NewGlobalRef(env).release();
```
Comment (ivovandongen) [inline platform/android/src/geometry/feature.cpp]: Thanks. Saw your comments too late. Will fix this up in a follow-up pr

File: platform/android/src/annotation/marker.cpp [hunk 2]
```diff
@@ -0,0 +1,31 @@
+#include "marker.hpp"
+
+namespace mbgl {
+namespace android {
+
+jni::Class<Marker> Marker::javaClass;
+
+mbgl::Point<double> Marker::getPosition(jni::JNIEnv& env, jni::Object<Marker> marker) {
+    static auto positionField = Marker::javaClass.GetField<jni::Object<LatLng>>(env, "position");
```
Comment (ivovandongen) [inline platform/android/src/annotation/marker.cpp]: There is one class definition for all threads as far as I know. The passed in env is assumed to be already attached.

**Pull request merged**